Optimize NaN-to-Number Comparisons

Geotale opened the pull request:
### Resolves

https://github.com/TurboWarp/scratch-vm/issues/72

### Proposed Changes

Optimizes Number-or-NaN to Number comparisons for greater and less than, by removing functions calls and either using operators normally when possible, or negating greater-equal or less-than-equal operators to effectively "switch the sides" that the NaN is on to make the operation remain correct.

### Reason for Changes

When these are not done then the `compareLessThan` and `compareGreaterThan` functions are used when they can be avoided, going to their slow counterparts if the operand in question actually is NaN.

### Test Coverage

Setting a variable to 0 / 0, then comparing it to a variety of values, including itself, and all permutations of constants, along with testing if these results matched up with those when disabling the compiler.

Geotale committed:
Optimize NaN-to-Number Comparisons

Optimizes Number-or-NaN to Number comparisons for greater and less than, by removing functions calls and either using operators normally when possible, or negating greater-equal or less-than-equal operators to effectively "switch the sides" that the NaN is on to make the operation remain correct.

GarboMuffin commented:
Will look at compatibility & measurable performance implications when I have time (probably next week)

hello-smile6 commented on src/compiler/jsgen.js:
```diff
@@ -559,11 +567,19 @@ class JSGenerator {
+                return new TypedInput(`!(${left.asNumber()} >= ${right.asNumberOrNaN()})`, TYPE_BOOLEAN);
+            }
             // When both operands are known to never be numbers, only use string comparison to avoid all number parsing.
             if (left.isNeverNumber() || right.isNeverNumber()) {
```
Shouldn't you use `&&`?

GarboMuffin replied:
No, the logic is correct and that line isn't even changed by this PR. (Although the comment is wrong)

Geotale replied:
If either are never numbers then I believe it should be safe to convert them to strings and compare them, as non-numbers can't really do weird things like `"0.000123" == "1.23e-4"`

hello-smile6 replied:
Maybe try `parseInt()` and assume values are not a number if an error is thrown?

GarboMuffin commented:
(I'm waiting until after the next desktop app update this week/next weekend just because I consider a change like this to be high risk for some edge case regression)

GarboMuffin committed:
Remove now-redundant compile-time comparison check

This condition is always false because of the earlier check:
left.isAlwaysNumber() && right.isAlwaysNumberOrNaN()

GarboMuffin commented:
tried awfully hard to find a compatibility issue but couldn't find any, let's just merge it and see if anything breaks

(probably won't make it into production for another week as things are *busy*)
